File: src/game/deck.cpp
// Copyright (c) 2022 r3w0p
// The following code can be redistributed and/or
// modified under the terms of the GPL-3.0 License.

#include <chrono>
#include <random>
#include "deck.h"
#include "exceptions.h"


Deck *DeckBuilder::build_caravan_deck(
        uint8_t num_cards,
        uint8_t num_sample_decks,
        bool balanced_sample) {

    uint8_t total_sample_cards;
    Deck sample_decks[num_sample_decks];
    Deck *d;
    uint8_t i_next;
    Card c_next;
    uint8_t first_hand_num_cards;

    if (num_cards < DECK_CARAVAN_MIN or
        num_cards > DECK_CARAVAN_MAX)
        throw CaravanFatalException(
                "A caravan deck must have between "
                "30 and 156 cards (inclusive).");

    if (num_sample_decks < SAMPLE_DECKS_MIN or
        num_sample_decks > SAMPLE_DECKS_MAX)
        throw CaravanFatalException(
                "A caravan deck must sample from between "
                "1 and 3 standard card decks (inclusive).");

    total_sample_cards = num_sample_decks * DECK_STANDARD_MAX;

    if (total_sample_cards < num_cards)
        throw CaravanFatalException(
                "There are insufficient cards to sample for the "
                "caravan deck.");

    d = new Deck();

    do {
        d->clear();
        i_next = 0;
        first_hand_num_cards = 0;

        for (int i = 0; i < num_sample_decks; ++i)
            sample_decks[i] = DeckBuilder::build_standard_deck(true);

        if (balanced_sample) {
            // Sample decks in a round-robin fashion
            while (d->size() < num_cards) {
                c_next = sample_decks[i_next].back();
                d->push_back(c_next);
                sample_decks[i_next].pop_back();

                i_next = (i_next + 1) % num_sample_decks;

                if ((num_cards - d->size()) < HAND_SIZE_MAX and
                    is_numeric_card(c_next))
                    first_hand_num_cards += 1;
            }

        } else {
            // Sample decks randomly
            std::random_device rd;
            std::mt19937 gen(rd());
            std::uniform_int_distribution<> distr(
                    0, num_sample_decks - 1);

            while (d->size() < num_cards) {
                i_next = distr(gen);
                if (!sample_decks[i_next].empty()) {
                    c_next = sample_decks[i_next].back();
                    d->push_back(c_next);
                    sample_decks[i_next].pop_back();

                    if ((num_cards - d->size()) < HAND_SIZE_MAX and
                        is_numeric_card(c_next))
                        first_hand_num_cards += 1;
                }
            }
        }

    } while (first_hand_num_cards < MOVES_START_ROUND);

    return d;
}

/*
 * PROTECTED
 */

Deck DeckBuilder::build_standard_deck(bool shuffle) {
    Deck d;

    for (int i = CLUBS; i <= SPADES; ++i)
        for (int j = ACE; j <= KING; ++j)
            d.push_back({
                static_cast<Suit>(i),
                static_cast<Rank>(j)
            });

    d.push_back({NO_SUIT, JOKER});
    d.push_back({NO_SUIT, JOKER});

    if (shuffle)
        return shuffle_deck(d);
    else
        return d;
}

Deck DeckBuilder::shuffle_deck(Deck d) {
    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
    std::shuffle(d.begin(), d.end(),
                 std::default_random_engine(seed));
    return d;
}

File: src/deck.rs
use rand::thread_rng;
use rand::seq::SliceRandom;

const DECK_STANDARD_MAX: u8 = 52;
const DECK_CARAVAN_MIN: u8 = 30;
const DECK_CARAVAN_MAX: u8 = 156;
const SAMPLE_DECKS_MIN: u8 = 1;
const SAMPLE_DECKS_MAX: u8 = 3;

#[derive(Debug)]
pub enum DeckError {
    NumCards, NumSamples
}

#[derive(Clone, Copy)]
pub enum Suit {
    NoSuit, Clubs, Diamonds, Hearts, Spades
}

impl Suit {
    pub fn to_str(&self) -> &str {
        match &self {
            Suit::NoSuit => "",
            Suit::Clubs => "CLUBS",
            Suit::Diamonds => "DIAMONDS",
            Suit::Hearts => "HEARTS",
            Suit::Spades => "SPADES"
        }
    }
}

pub enum Rank {
    Ace, Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten,
    Jack, Queen, King, Joker
}

impl Rank {
    pub fn to_str(&self) -> &str {
        match &self {
            Rank::Ace => "ACE",
            Rank::Two => "TWO",
            Rank::Three => "THREE",
            Rank::Four => "FOUR",
            Rank::Five => "FIVE",
            Rank::Six => "SIX",
            Rank::Seven => "SEVEN",
            Rank::Eight => "EIGHT",
            Rank::Nine => "NINE",
            Rank::Ten => "TEN",
            Rank::Jack => "JACK",
            Rank::Queen => "QUEEN",
            Rank::King => "KING",
            Rank::Joker => "JOKER",
        }
    }
}

pub struct Card {
    pub suit: Suit,
    pub rank: Rank
}

pub type Deck = Vec<Card>;

pub fn build_caravan_deck(
    num_cards: u8,
    num_sample_decks: u8,
    balanced: bool) -> Result<Deck, DeckError> {

    let total_sample_cards: u8;
    let sample_decks: [Deck; SAMPLE_DECKS_MAX as usize] = [];
    
    if num_cards < DECK_CARAVAN_MIN { Err(DeckError::NumCards) }
    if num_cards > DECK_CARAVAN_MAX { Err(DeckError::NumCards) }

    if num_sample_decks < SAMPLE_DECKS_MIN { Err(DeckError::NumSamples) }
    if num_sample_decks > SAMPLE_DECKS_MAX { Err(DeckError::NumSamples) }

    total_sample_cards = num_sample_decks * DECK_STANDARD_MAX;

    if total_sample_cards < num_cards { Err(DeckError::NumCards) }

    for i in 0..num_sample_decks {
        sample_decks[i] = build_standard_deck(true);
    }

    if balanced {

    } else {

    }

    Ok(build_standard_deck(false))  // TODO
}

fn build_standard_deck(shuffle: bool) -> Deck {
    let mut d: Deck = Deck::new();

    for s in [
        Suit::Clubs,
        Suit::Diamonds,
        Suit::Hearts,
        Suit::Spades
    ] {
        for r in [
            Rank::Ace,
            Rank::Two,
            Rank::Three,
            Rank::Four,
            Rank::Five,
            Rank::Six,
            Rank::Seven,
            Rank::Eight,
            Rank::Nine,
            Rank::Ten,
            Rank::Jack,
            Rank::Queen,
            Rank::King,
        ] {
            d.push(Card {
                suit: s,
                rank: r
            });
        }
    }

    for _ in 0..2 {
        d.push(Card {
            suit: Suit::NoSuit,
            rank: Rank::Joker
        });
    }

    if shuffle {
        d.shuffle(&mut thread_rng());
    }

    d
}
